File: src/cadforge/core/session.py
"""Session persistence for CadForge.

Stores conversation transcripts as JSONL files with a session index
for resume and search.
"""

from __future__ import annotations

import json
import time
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from cadforge.utils.paths import get_project_sessions_dir


@dataclass
class Message:
    """A single conversation message."""
    role: str  # "user", "assistant", "system"
    content: str | list[dict[str, Any]]
    timestamp: float = field(default_factory=time.time)
    tool_use_id: str | None = None
    tool_result: Any = None
    usage: dict[str, int] | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "role": self.role,
            "content": self.content,
            "timestamp": self.timestamp,
        }
        if self.tool_use_id:
            d["tool_use_id"] = self.tool_use_id
        if self.tool_result is not None:
            d["tool_result"] = self.tool_result
        if self.usage:
            d["usage"] = self.usage
        return d

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Message:
        return cls(
            role=d["role"],
            content=d["content"],
            timestamp=d.get("timestamp", 0.0),
            tool_use_id=d.get("tool_use_id"),
            tool_result=d.get("tool_result"),
            usage=d.get("usage"),
        )


@dataclass
class SessionMetadata:
    """Metadata about a session for the index."""
    session_id: str
    started_at: str
    last_active: str
    message_count: int
    summary: str
    file_path: str

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> SessionMetadata:
        return cls(**d)


class Session:
    """Manages a single conversation session with JSONL persistence."""

    def __init__(self, project_root: Path, session_id: str | None = None):
        self.project_root = project_root
        self.session_id = session_id or self._generate_id()
        self.sessions_dir = get_project_sessions_dir(project_root)
        self.file_path = self.sessions_dir / f"{self.session_id}.jsonl"
        self.messages: list[Message] = []
        self._loaded = False

    @staticmethod
    def _generate_id() -> str:
        return datetime.now(timezone.utc).strftime("session-%Y-%m-%d-%H-%M-%S")

    def add_message(self, message: Message) -> None:
        """Add a message and persist to disk."""
        self.messages.append(message)
        self._append_to_file(message)

    def add_user_message(self, content: str) -> Message:
        """Convenience method for adding a user message."""
        msg = Message(role="user", content=content)
        self.add_message(msg)
        return msg

    def add_assistant_message(
        self,
        content: str | list[dict[str, Any]],
        usage: dict[str, int] | None = None,
    ) -> Message:
        """Convenience method for adding an assistant message."""
        msg = Message(role="assistant", content=content, usage=usage)
        self.add_message(msg)
        return msg

    def _append_to_file(self, message: Message) -> None:
        """Append a single message to the JSONL file."""
        self.file_path.parent.mkdir(parents=True, exist_ok=True)
        with open(self.file_path, "a", encoding="utf-8") as f:
            f.write(json.dumps(message.to_dict()) + "\n")

    def load(self) -> None:
        """Load messages from JSONL file."""
        if not self.file_path.exists():
            return
        self.messages = []
        with open(self.file_path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line:
                    self.messages.append(Message.from_dict(json.loads(line)))
        self._loaded = True

    def get_api_messages(self) -> list[dict[str, Any]]:
        """Get messages formatted for the Anthropic API."""
        api_msgs = []
        for msg in self.messages:
            if msg.role in ("user", "assistant"):
                api_msgs.append({
                    "role": msg.role,
                    "content": msg.content,
                })
        return api_msgs

    @property
    def message_count(self) -> int:
        return len(self.messages)

    @property
    def summary(self) -> str:
        """Generate a brief summary from the first user message."""
        for msg in self.messages:
            if msg.role == "user" and isinstance(msg.content, str):
                text = msg.content[:100]
                if len(msg.content) > 100:
                    text += "..."
                return text
        return "(empty session)"

    def to_metadata(self) -> SessionMetadata:
        """Generate metadata for the session index."""
        started = self.messages[0].timestamp if self.messages else time.time()
        last = self.messages[-1].timestamp if self.messages else started
        return SessionMetadata(
            session_id=self.session_id,
            started_at=datetime.fromtimestamp(started, timezone.utc).isoformat(),
            last_active=datetime.fromtimestamp(last, timezone.utc).isoformat(),
            message_count=self.message_count,
            summary=self.summary,
            file_path=str(self.file_path),
        )


class SessionIndex:
    """Manages the session index file for listing and resuming sessions."""

    def __init__(self, project_root: Path):
        self.project_root = project_root
        self.sessions_dir = get_project_sessions_dir(project_root)
        self.index_path = self.sessions_dir / "sessions-index.json"

    def _load_index(self) -> list[SessionMetadata]:
        if not self.index_path.exists():
            return []
        try:
            data = json.loads(self.index_path.read_text(encoding="utf-8"))
            return [SessionMetadata.from_dict(d) for d in data]
        except (json.JSONDecodeError, KeyError):
            return []

    def _save_index(self, entries: list[SessionMetadata]) -> None:
        self.index_path.parent.mkdir(parents=True, exist_ok=True)
        self.index_path.write_text(
            json.dumps([e.to_dict() for e in entries], indent=2) + "\n",
            encoding="utf-8",
        )

    def update(self, session: Session) -> None:
        """Update or add a session in the index."""
        entries = self._load_index()
        metadata = session.to_metadata()

        # Replace existing entry or append
        for i, entry in enumerate(entries):
            if entry.session_id == metadata.session_id:
                entries[i] = metadata
                break
        else:
            entries.append(metadata)

        self._save_index(entries)

    def list_sessions(self, limit: int = 20) -> list[SessionMetadata]:
        """List recent sessions, most recent first."""
        entries = self._load_index()
        entries.sort(key=lambda e: e.last_active, reverse=True)
        return entries[:limit]

    def get_latest(self) -> SessionMetadata | None:
        """Get the most recent session."""
        entries = self.list_sessions(limit=1)
        return entries[0] if entries else None

    def get_by_id(self, session_id: str) -> SessionMetadata | None:
        """Find a session by its ID."""
        entries = self._load_index()
        for entry in entries:
            if entry.session_id == session_id:
                return entry
        return None

File: packages/cli/src/session/session.ts
/**
 * Session persistence — JSONL read/write (same format as Python).
 *
 * Stores conversation transcripts as JSONL files with a session index
 * for resume and search.
 */

import { appendFileSync, existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { getCadforgeDir } from '../config/paths.js';

export interface SessionEntry {
  role: 'user' | 'assistant' | 'system';
  content: string | Record<string, unknown>[];
  timestamp: number;
  tool_use_id?: string;
  tool_result?: unknown;
  usage?: { input_tokens: number; output_tokens: number };
}

export interface SessionMetadata {
  session_id: string;
  started_at: string;
  last_active: string;
  message_count: number;
  summary: string;
  file_path: string;
}

function getSessionsDir(projectRoot: string): string {
  const d = join(getCadforgeDir(projectRoot), 'sessions');
  mkdirSync(d, { recursive: true });
  return d;
}

function generateSessionId(): string {
  const now = new Date();
  const pad = (n: number, len = 2) => String(n).padStart(len, '0');
  return `session-${now.getUTCFullYear()}-${pad(now.getUTCMonth() + 1)}-${pad(now.getUTCDate())}-${pad(now.getUTCHours())}-${pad(now.getUTCMinutes())}-${pad(now.getUTCSeconds())}`;
}

export class Session {
  readonly id: string;
  readonly projectRoot: string;
  readonly filePath: string;
  readonly entries: SessionEntry[] = [];

  constructor(projectRoot: string, id?: string) {
    this.projectRoot = projectRoot;
    this.id = id ?? generateSessionId();
    this.filePath = join(getSessionsDir(projectRoot), `${this.id}.jsonl`);
  }

  addMessage(entry: SessionEntry): void {
    this.entries.push(entry);
    this.appendToFile(entry);
  }

  addUserMessage(content: string | Record<string, unknown>[]): SessionEntry {
    const entry: SessionEntry = { role: 'user', content, timestamp: Date.now() / 1000 };
    this.addMessage(entry);
    return entry;
  }

  addAssistantMessage(
    content: string | Record<string, unknown>[],
    usage?: { input_tokens: number; output_tokens: number },
  ): SessionEntry {
    const entry: SessionEntry = { role: 'assistant', content, timestamp: Date.now() / 1000 };
    if (usage) entry.usage = usage;
    this.addMessage(entry);
    return entry;
  }

  private appendToFile(entry: SessionEntry): void {
    const dir = dirname(this.filePath);
    mkdirSync(dir, { recursive: true });
    const data: Record<string, unknown> = {
      role: entry.role,
      content: entry.content,
      timestamp: entry.timestamp,
    };
    if (entry.tool_use_id) data.tool_use_id = entry.tool_use_id;
    if (entry.tool_result !== undefined) data.tool_result = entry.tool_result;
    if (entry.usage) data.usage = entry.usage;
    appendFileSync(this.filePath, JSON.stringify(data) + '\n', 'utf-8');
  }

  load(): void {
    if (!existsSync(this.filePath)) return;
    const text = readFileSync(this.filePath, 'utf-8');
    this.entries.length = 0;
    for (const line of text.split('\n')) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      const d = JSON.parse(trimmed) as Record<string, unknown>;
      this.entries.push({
        role: d.role as SessionEntry['role'],
        content: d.content as SessionEntry['content'],
        timestamp: (d.timestamp as number) ?? 0,
        tool_use_id: d.tool_use_id as string | undefined,
        tool_result: d.tool_result,
        usage: d.usage as SessionEntry['usage'],
      });
    }
  }

  /**
   * Get messages formatted for the Anthropic API.
   * Handles dangling tool_use blocks by injecting error tool_result blocks.
   */
  getApiMessages(): Record<string, unknown>[] {
    const apiMsgs: Record<string, unknown>[] = [];
    for (const entry of this.entries) {
      if (entry.role === 'user' || entry.role === 'assistant') {
        apiMsgs.push({ role: entry.role, content: entry.content });
      }
    }

    // Fix dangling tool_use
    if (apiMsgs.length > 0 && apiMsgs[apiMsgs.length - 1].role === 'assistant') {
      const content = apiMsgs[apiMsgs.length - 1].content;
      if (Array.isArray(content)) {
        const toolUseIds = (content as Record<string, unknown>[])
          .filter((b) => b.type === 'tool_use')
          .map((b) => b.id as string);
        if (toolUseIds.length > 0) {
          const errorResults = toolUseIds.map((id) => ({
            type: 'tool_result',
            tool_use_id: id,
            content: 'Session was interrupted before tool execution completed.',
            is_error: true,
          }));
          apiMsgs.push({ role: 'user', content: errorResults });
        }
      }
    }

    return apiMsgs;
  }

  get messageCount(): number {
    return this.entries.length;
  }

  get summary(): string {
    for (const entry of this.entries) {
      if (entry.role === 'user' && typeof entry.content === 'string') {
        const text = entry.content.slice(0, 100);
        return entry.content.length > 100 ? text + '...' : text;
      }
    }
    return '(empty session)';
  }

  toMetadata(): SessionMetadata {
    const started = this.entries.length > 0 ? this.entries[0].timestamp : Date.now() / 1000;
    const last = this.entries.length > 0 ? this.entries[this.entries.length - 1].timestamp : started;
    return {
      session_id: this.id,
      started_at: new Date(started * 1000).toISOString(),
      last_active: new Date(last * 1000).toISOString(),
      message_count: this.messageCount,
      summary: this.summary,
      file_path: this.filePath,
    };
  }
}

export class SessionIndex {
  private readonly indexPath: string;

  constructor(projectRoot: string) {
    const sessionsDir = getSessionsDir(projectRoot);
    this.indexPath = join(sessionsDir, 'sessions-index.json');
  }

  private loadIndex(): SessionMetadata[] {
    if (!existsSync(this.indexPath)) return [];
    try {
      return JSON.parse(readFileSync(this.indexPath, 'utf-8')) as SessionMetadata[];
    } catch {
      return [];
    }
  }

  private saveIndex(entries: SessionMetadata[]): void {
    const dir = dirname(this.indexPath);
    mkdirSync(dir, { recursive: true });
    writeFileSync(this.indexPath, JSON.stringify(entries, null, 2) + '\n', 'utf-8');
  }

  update(session: Session): void {
    const entries = this.loadIndex();
    const metadata = session.toMetadata();
    const idx = entries.findIndex((e) => e.session_id === metadata.session_id);
    if (idx >= 0) {
      entries[idx] = metadata;
    } else {
      entries.push(metadata);
    }
    this.saveIndex(entries);
  }

  listSessions(limit = 20): SessionMetadata[] {
    const entries = this.loadIndex();
    entries.sort((a, b) => b.last_active.localeCompare(a.last_active));
    return entries.slice(0, limit);
  }

  getLatest(): SessionMetadata | null {
    const entries = this.listSessions(1);
    return entries.length > 0 ? entries[0] : null;
  }

  getById(sessionId: string): SessionMetadata | null {
    const entries = this.loadIndex();
    return entries.find((e) => e.session_id === sessionId) ?? null;
  }
}
